### PR TITLE
[SPARK-41427][UI] Protobuf serializer for ExecutorStageSummaryWrapper

### PR DESCRIPTION
**Pull request by gengliangwang**
### What changes were proposed in this pull request?
Add Protobuf serializer for ExecutorStageSummary

### Why are the changes needed?
Support fast and compact serialization/deserialization for ExecutorStageSummary over RocksDB.

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
New UT

### CODE REVIEW

**Comment by gengliangwang** [comment]
cc @techaddict

File: core/src/main/scala/org/apache/spark/status/protobuf/ExecutorMetricsSerializer.scala [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.spark.status.protobuf
+
+import org.apache.spark.executor.ExecutorMetrics
+import org.apache.spark.metrics.ExecutorMetricType
+
+object ExecutorMetricsSerializer {
+  def serialize(e: ExecutorMetrics): StoreTypes.ExecutorMetrics = {
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/protobuf/ExecutorMetricsSerializer.scala:24]: This follows ExecutorMetricsJsonSerializer

File: core/src/main/scala/org/apache/spark/status/protobuf/ExecutorMetricsSerializer.scala [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    val builder = StoreTypes.ExecutorMetrics.newBuilder()
+    ExecutorMetricType.metricToOffset.foreach { case (metric, _) =>
+      builder.putMetrics(metric, e.getMetricValue(metric))
+      metric -> e.getMetricValue(metric)
+    }
+    builder.build()
+  }
+
+  def deserialize(binary: StoreTypes.ExecutorMetrics): ExecutorMetrics = {
```
Comment (gengliangwang) [inline core/src/main/scala/org/apache/spark/status/protobuf/ExecutorMetricsSerializer.scala:33]: This follows ExecutorMetricsJsonDeserializer

**Comment by techaddict** [comment]
@gengliangwang is this not for https://issues.apache.org/jira/browse/SPARK-41427 ?
This one looks like PR for `ExecutorStageSummaryWrapper` where as issue [SPARK-41422] was for `ExecutorSummaryWrapper`

**Comment by gengliangwang** [comment]
@techaddict Yes it is for ExecutorStageSummaryWrapper. I just updated the title.

**Comment by LuciferYang** [comment]
hmm.. is this should be SPARK-41427？

**Review by cloud-fan — APPROVED** [review]

**Comment by gengliangwang** [comment]
Merging to master

File: core/src/main/scala/org/apache/spark/status/protobuf/ExecutorMetricsSerializer.scala [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+import org.apache.spark.executor.ExecutorMetrics
+import org.apache.spark.metrics.ExecutorMetricType
+
+object ExecutorMetricsSerializer {
+  def serialize(e: ExecutorMetrics): StoreTypes.ExecutorMetrics = {
+    val builder = StoreTypes.ExecutorMetrics.newBuilder()
+    ExecutorMetricType.metricToOffset.foreach { case (metric, _) =>
+      builder.putMetrics(metric, e.getMetricValue(metric))
+      metric -> e.getMetricValue(metric)
```
Comment (LuciferYang) [inline core/src/main/scala/org/apache/spark/status/protobuf/ExecutorMetricsSerializer.scala:28]: @gengliangwang 
line 27 already `putMetrics(metric, e.getMetricValue(metric))`,

What is the purpose of line 28 (`metric -> e.getMetricValue(metric)` )? I didn't understand it. Could you explain? Thanks.

Comment (LuciferYang) [reply core/src/main/scala/org/apache/spark/status/protobuf/ExecutorMetricsSerializer.scala:28]: I guess it's redundant, so I give https://github.com/apache/spark/pull/39135 to delete line 28, if not, please correct me

**Pull request closed without merging**